### PR TITLE
Editor embeds merge in nitro embed

### LINKED ISSUE #256: [Editor] Embeds
> [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-01-07T16:08:19Z" title="Monday, January 7th 2019, 5:08:19 pm +01:00">Jan 7, 2019</time>_
_Closed <time datetime="2019-08-06T05:50:10Z" title="Tuesday, August 6th 2019, 7:50:10 am +02:00">Aug 6, 2019</time>_
---

As a user
I would like to see link previews and video embeds
In order to not have to click or leave the page to get the contents
---
We need to render `a` tags with a `data-url-embed` attribute an embed component. As a reference we have to look into the implementation of the todo-item node from the tiptap examples.
https://github.com/scrumpy/tiptap/blob/master/packages/tiptap-extensions/src/nodes/TodoItem.js
https://tiptap.scrumpy.io/todo-list

### PR DESCRIPTION
**Pull request by Tirokk**
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-03T11:53:33Z" title="Wednesday, July 3rd 2019, 1:53:33 pm +02:00">Jul 3, 2019</time>_
_Merged <time datetime="2019-07-22T12:02:37Z" title="Monday, July 22nd 2019, 2:02:37 pm +02:00">Jul 22, 2019</time>_
---

### Issues

- fixes #256

### CODE REVIEW

**Comment by Tirokk** [comment]
> [codecov[bot]](https://github.com/apps/codecov) **Authored by [codecov[bot]](https://github.com/apps/codecov)**
_<time datetime="2019-07-03T14:13:52Z" title="Wednesday, July 3rd 2019, 4:13:52 pm +02:00">Jul 3, 2019</time>_

---


# [Codecov](https://codecov.io/gh/Human-Connection/Human-Connection/pull/960?src=pr&el=h1) Report
> :exclamation: No coverage uploaded for pull request base (`master@324330a`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> The diff coverage is `n/a`.

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-12T10:25:56Z" title="Friday, July 12th 2019, 12:25:56 pm +02:00">Jul 12, 2019</time>_

---


does this look ok @roschaefer? I used `tomono`, but because I already had a repo with the name `Human-Connection`, I needed to remove it and recreate it with the merged commit history. This caused some small issues for me locally, but I think others won't face the same.

The files changed and commit history look good to me

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-07-16T19:44:24Z" title="Tuesday, July 16th 2019, 9:44:24 pm +02:00">Jul 16, 2019</time>_

---


@mattwr18 this looks good, but could sb. explain me why this is a separate service and not just another resolver in our backend?

@appinteractive maybe?

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-16T19:48:31Z" title="Tuesday, July 16th 2019, 9:48:31 pm +02:00">Jul 16, 2019</time>_

---


I think it could be, I thought the same thing myself @roschaefer

File: backend/src/schema/resolvers/embeds.spec.js [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+              title: 'Baby Loves Cat',
+              author: 'Merkley Family',
+              publisher: 'YouTube',
+              date: expect.any(String),
+              description:
+                'She’s incapable of controlling her limbs when her kitty is around. The obsession grows every day. Ps. That’s a sleep sack she’s in. Not a starfish outfit. Al...',
+              url: 'https://www.youtube.com/watch?v=qkdXAtO40Fo',
+              image: 'https://i.ytimg.com/vi/qkdXAtO40Fo/maxresdefault.jpg',
+              audio: null,
+              video: null,
```
Comment (Tirokk) [inline backend/src/schema/resolvers/embeds.spec.js:204]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-07-21T22:41:11Z" title="Monday, July 22nd 2019, 12:41:11 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,216 @@
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../../server'
+import { gql } from '../../jest/helpers'
+
+jest.mock('node-fetch')
+const { Response } = jest.requireActual('node-fetch')
+
+afterEach(() => {
+  fetch.mockRestore()
+})
+
+let variables = {}
+
+const HumanConnectionOrg = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/HumanConnectionOrg.html'),
+  'utf8',
+)
+const pr960 = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/pr960.html'),
+  'utf8',
+)
+const babyLovesCat = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/babyLovesCat.html'),
+  'utf8',
+)
+
+const babyLovesCatEmbedResponse = new Response(
+  JSON.stringify({
+    height: 270,
+    provider_name: 'YouTube',
+    title: 'Baby Loves Cat',
+    type: 'video',
+    width: 480,
+    thumbnail_height: 360,
+    provider_url: 'https://www.youtube.com/',
+    thumbnail_width: 480,
+    html:
+      '<iframe width="480" height="270" src="https://www.youtube.com/embed/qkdXAtO40Fo?start=18&feature=oembed" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>',
+    thumbnail_url: 'https://i.ytimg.com/vi/qkdXAtO40Fo/hqdefault.jpg',
+    version: '1.0',
+    author_name: 'Merkley Family',
+    author_url: 'https://www.youtube.com/channel/UC5P8yei950tif7UmdPpkJLQ',
+  }),
+)
+
+describe('Query', () => {
+  describe('embed', () => {
+    let embedAction
+
+    beforeEach(() => {
+      embedAction = async variables => {
+        const { server } = createServer({
+          context: () => {},
+        })
+        const { query } = createTestClient(server)
+        const embed = gql`
+          query($url: String!) {
+            embed(url: $url) {
+              type
+              title
+              author
+              publisher
+              date
+              description
+              url
+              image
+              audio
+              video
+              lang
+              sources
+              html
+            }
+          }
+        `
+        return query({ query: embed, variables })
+      }
+    })
+
+    describe('given a video link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response('')))
+          .mockReturnValueOnce(Promise.resolve(JSON.stringify({})))
+        variables = { url: 'https://www.w3schools.com/html/mov_bbb.mp4' }
+      })
+
+      it('shows some default data', async () => {
+        const expected = expect.objectContaining({
+          data: {
+            embed: {
+              audio: null,
+              author: null,
+              date: null,
+              description: null,
+              html: null,
+              image: null,
+              lang: null,
+              publisher: null,
+              sources: ['resource'],
+              title: null,
+              type: 'link',
+              url: 'https://www.w3schools.com/html/mov_bbb.mp4',
+              video: null,
+            },
+          },
+        })
+        await expect(embedAction(variables)).resolves.toEqual(expected)
+      })
+    })
+
+    describe('given a Facebook link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response(HumanConnectionOrg)))
+          .mockReturnValueOnce(Promise.resolve('invalid json'))
+        variables = { url: 'https://www.facebook.com/HumanConnectionOrg/' }
+      })
+
+      it('does not crash if embed provider returns invalid JSON', async () => {
+        const expected = expect.objectContaining({
+          data: {
+            embed: {
+              audio: null,
+              author: null,
+              date: expect.any(String),
+              description:
+                'Human Connection, Weilheim an der Teck. Gefällt 24.407 Mal. An upcoming non-profit social network focused on local and global positive change. Twitter accounts : @hc_world (EN), @hc_deutschland (GE),...',
+              html: null,
+              image:
+                'https://scontent.ftxl3-1.fna.fbcdn.net/v/t1.0-1/c5.0.200.200a/p200x200/12108307_997373093648222_70057205881020137_n.jpg?_nc_cat=110&_nc_oc=AQnPPYQlR0dU556gOfl4xkXr7IPZdRIAUfQeXl3fpUv4DAsFN8T4PfgOjPwuq85GPKGZ5S5E5mWQ8IVV1UiRBAIZ&_nc_ht=scontent.ftxl3-1.fna&oh=90309adddaab38839782f16e7d4b7bcf&oe=5DEEDFE5',
+              lang: 'de',
+              publisher: 'Facebook',
+              sources: ['resource'],
+              title: 'Human Connection',
+              type: 'link',
+              url: 'https://www.facebook.com/HumanConnectionOrg/',
+              video: null,
+            },
+          },
+        })
+        await expect(embedAction(variables)).resolves.toEqual(expected)
+      })
+    })
+
+    describe('given a Github link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response(pr960)))
+          .mockReturnValueOnce(Promise.resolve(JSON.stringify({})))
+        variables = { url: 'https://github.com/Human-Connection/Human-Connection/pull/960' }
+      })
+
+      it('returns meta data even if no embed html can be retrieved', async () => {
+        const expected = expect.objectContaining({
+          data: {
+            embed: {
+              type: 'link',
+              title:
+                'Editor embeds merge in nitro embed by mattwr18 · Pull Request #960 · Human-Connection/Human-Connection',
+              author: 'Human-Connection',
+              publisher: 'GitHub',
+              date: expect.any(String),
+              description: '🍰 Pullrequest Issues fixes #256',
+              url: 'https://github.com/Human-Connection/Human-Connection/pull/960',
+              image:
+                'https://repository-images.githubusercontent.com/112590397/52c9a000-7e11-11e9-899d-aaa55f3a3d72',
+              audio: null,
+              video: null,
+              lang: 'en',
+              sources: ['resource'],
+              html: null,
+            },
+          },
+        })
+        await expect(embedAction(variables)).resolves.toEqual(expected)
+      })
+    })
+
+    describe('given a youtube link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response(babyLovesCat)))
+          .mockReturnValueOnce(Promise.resolve(babyLovesCatEmbedResponse))
+        variables = { url: 'https://www.youtube.com/watch?v=qkdXAtO40Fo&t=18s' }
+      })
+
+      it('returns meta data plus youtube iframe html', async () => {
+        const expected = expect.objectContaining({
+          data: {
+            embed: {
+              type: 'video',
+              title: 'Baby Loves Cat',
+              author: 'Merkley Family',
+              publisher: 'YouTube',
+              date: expect.any(String),
+              description:
+                'She’s incapable of controlling her limbs when her kitty is around. The obsession grows every day. Ps. That’s a sleep sack she’s in. Not a starfish outfit. Al...',
+              url: 'https://www.youtube.com/watch?v=qkdXAtO40Fo',
+              image: 'https://i.ytimg.com/vi/qkdXAtO40Fo/maxresdefault.jpg',
+              audio: null,
+              video: null,
```


`audio` and `video` was always `null` in all my tests...

File: backend/src/schema/resolvers/embeds.spec.js [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    author_url: 'https://www.youtube.com/channel/UC5P8yei950tif7UmdPpkJLQ',
+  }),
+)
+
+describe('Query', () => {
+  describe('embed', () => {
+    let embedAction
+
+    beforeEach(() => {
+      embedAction = async variables => {
```
Comment (Tirokk) [inline backend/src/schema/resolvers/embeds.spec.js:54]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:26:41Z" title="Monday, July 22nd 2019, 9:26:41 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,216 @@
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../../server'
+import { gql } from '../../jest/helpers'
+
+jest.mock('node-fetch')
+const { Response } = jest.requireActual('node-fetch')
+
+afterEach(() => {
+  fetch.mockRestore()
+})
+
+let variables = {}
+
+const HumanConnectionOrg = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/HumanConnectionOrg.html'),
+  'utf8',
+)
+const pr960 = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/pr960.html'),
+  'utf8',
+)
+const babyLovesCat = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/babyLovesCat.html'),
+  'utf8',
+)
+
+const babyLovesCatEmbedResponse = new Response(
+  JSON.stringify({
+    height: 270,
+    provider_name: 'YouTube',
+    title: 'Baby Loves Cat',
+    type: 'video',
+    width: 480,
+    thumbnail_height: 360,
+    provider_url: 'https://www.youtube.com/',
+    thumbnail_width: 480,
+    html:
+      '<iframe width="480" height="270" src="https://www.youtube.com/embed/qkdXAtO40Fo?start=18&feature=oembed" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>',
+    thumbnail_url: 'https://i.ytimg.com/vi/qkdXAtO40Fo/hqdefault.jpg',
+    version: '1.0',
+    author_name: 'Merkley Family',
+    author_url: 'https://www.youtube.com/channel/UC5P8yei950tif7UmdPpkJLQ',
+  }),
+)
+
+describe('Query', () => {
+  describe('embed', () => {
+    let embedAction
+
+    beforeEach(() => {
+      embedAction = async variables => {
```


:heart: the naming convention @roschaefer

File: backend/src/schema/resolvers/embeds.spec.js [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+    describe('given a Facebook link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response(HumanConnectionOrg)))
+          .mockReturnValueOnce(Promise.resolve('invalid json'))
+        variables = { url: 'https://www.facebook.com/HumanConnectionOrg/' }
+      })
+
+      it('does not crash if embed provider returns invalid JSON', async () => {
```
Comment (Tirokk) [inline backend/src/schema/resolvers/embeds.spec.js:122]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:28:06Z" title="Monday, July 22nd 2019, 9:28:06 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,216 @@
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+import { createTestClient } from 'apollo-server-testing'
+import createServer from '../../server'
+import { gql } from '../../jest/helpers'
+
+jest.mock('node-fetch')
+const { Response } = jest.requireActual('node-fetch')
+
+afterEach(() => {
+  fetch.mockRestore()
+})
+
+let variables = {}
+
+const HumanConnectionOrg = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/HumanConnectionOrg.html'),
+  'utf8',
+)
+const pr960 = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/pr960.html'),
+  'utf8',
+)
+const babyLovesCat = fs.readFileSync(
+  path.join(__dirname, '../../jest/snapshots/embeds/babyLovesCat.html'),
+  'utf8',
+)
+
+const babyLovesCatEmbedResponse = new Response(
+  JSON.stringify({
+    height: 270,
+    provider_name: 'YouTube',
+    title: 'Baby Loves Cat',
+    type: 'video',
+    width: 480,
+    thumbnail_height: 360,
+    provider_url: 'https://www.youtube.com/',
+    thumbnail_width: 480,
+    html:
+      '<iframe width="480" height="270" src="https://www.youtube.com/embed/qkdXAtO40Fo?start=18&feature=oembed" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>',
+    thumbnail_url: 'https://i.ytimg.com/vi/qkdXAtO40Fo/hqdefault.jpg',
+    version: '1.0',
+    author_name: 'Merkley Family',
+    author_url: 'https://www.youtube.com/channel/UC5P8yei950tif7UmdPpkJLQ',
+  }),
+)
+
+describe('Query', () => {
+  describe('embed', () => {
+    let embedAction
+
+    beforeEach(() => {
+      embedAction = async variables => {
+        const { server } = createServer({
+          context: () => {},
+        })
+        const { query } = createTestClient(server)
+        const embed = gql`
+          query($url: String!) {
+            embed(url: $url) {
+              type
+              title
+              author
+              publisher
+              date
+              description
+              url
+              image
+              audio
+              video
+              lang
+              sources
+              html
+            }
+          }
+        `
+        return query({ query: embed, variables })
+      }
+    })
+
+    describe('given a video link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response('')))
+          .mockReturnValueOnce(Promise.resolve(JSON.stringify({})))
+        variables = { url: 'https://www.w3schools.com/html/mov_bbb.mp4' }
+      })
+
+      it('shows some default data', async () => {
+        const expected = expect.objectContaining({
+          data: {
+            embed: {
+              audio: null,
+              author: null,
+              date: null,
+              description: null,
+              html: null,
+              image: null,
+              lang: null,
+              publisher: null,
+              sources: ['resource'],
+              title: null,
+              type: 'link',
+              url: 'https://www.w3schools.com/html/mov_bbb.mp4',
+              video: null,
+            },
+          },
+        })
+        await expect(embedAction(variables)).resolves.toEqual(expected)
+      })
+    })
+
+    describe('given a Facebook link', () => {
+      beforeEach(() => {
+        fetch
+          .mockReturnValueOnce(Promise.resolve(new Response(HumanConnectionOrg)))
+          .mockReturnValueOnce(Promise.resolve('invalid json'))
+        variables = { url: 'https://www.facebook.com/HumanConnectionOrg/' }
+      })
+
+      it('does not crash if embed provider returns invalid JSON', async () => {
```


Any idea when providers return invalid JSONs and why?

File: backend/src/schema/resolvers/embeds.spec.js (hunk 1 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/embeds.spec.js:204]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:28:28Z" title="Monday, July 22nd 2019, 9:28:28 am +02:00">Jul 22, 2019</time>_

---


strange, that was what I observed as well

File: backend/src/schema/resolvers/embeds/scraper.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+// some providers allow a format parameter
+// we need JSON
+oEmbedProvidersFile = oEmbedProvidersFile.replace('{format}', 'json')
+
+const oEmbedProviders = JSON.parse(oEmbedProvidersFile)
+
+const fetchEmbed = async url => {
+  const provider = oEmbedProviders.find(provider => {
+    return provider.provider_url.includes(url.hostname)
```
Comment (Tirokk) [inline backend/src/schema/resolvers/embeds/scraper.js:43]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:30:16Z" title="Monday, July 22nd 2019, 9:30:16 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,102 @@
+import Metascraper from 'metascraper'
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+
+import { ApolloError } from 'apollo-server'
+import isEmpty from 'lodash/isEmpty'
+import isArray from 'lodash/isArray'
+import mergeWith from 'lodash/mergeWith'
+
+const error = require('debug')('embed:error')
+
+const metascraper = Metascraper([
+  require('metascraper-author')(),
+  require('metascraper-date')(),
+  require('metascraper-description')(),
+  require('metascraper-image')(),
+  require('metascraper-lang')(),
+  require('metascraper-lang-detector')(),
+  require('metascraper-logo')(),
+  // require('metascraper-clearbit-logo')(),
+  require('metascraper-publisher')(),
+  require('metascraper-title')(),
+  require('metascraper-url')(),
+  require('metascraper-audio')(),
+  require('metascraper-soundcloud')(),
+  require('metascraper-video')(),
+  require('metascraper-youtube')(),
+
+  // require('./rules/metascraper-embed')()
+])
+
+let oEmbedProvidersFile = fs.readFileSync(path.join(__dirname, './providers.json'), 'utf8')
+
+// some providers allow a format parameter
+// we need JSON
+oEmbedProvidersFile = oEmbedProvidersFile.replace('{format}', 'json')
+
+const oEmbedProviders = JSON.parse(oEmbedProvidersFile)
+
+const fetchEmbed = async url => {
+  const provider = oEmbedProviders.find(provider => {
+    return provider.provider_url.includes(url.hostname)
```


can you explain where you got this long list of providers? how does this work actually? we added one of @Tirokk personal sites and it returned the metadata, but not sure how any of the hard-coded providers include his url hostname :thinking:

File: backend/src/schema/resolvers/embeds/scraper.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  return {
+    sources: ['resource'],
+    ...resource,
+  }
+}
+
+export default async function scrape(url) {
+  url = new URL(url)
+  if (url.hostname === 'youtu.be') {
+    // replace youtu.be to get proper results
```
Comment (Tirokk) [inline backend/src/schema/resolvers/embeds/scraper.js:83]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:30:37Z" title="Monday, July 22nd 2019, 9:30:37 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,102 @@
+import Metascraper from 'metascraper'
+import fetch from 'node-fetch'
+import fs from 'fs'
+import path from 'path'
+
+import { ApolloError } from 'apollo-server'
+import isEmpty from 'lodash/isEmpty'
+import isArray from 'lodash/isArray'
+import mergeWith from 'lodash/mergeWith'
+
+const error = require('debug')('embed:error')
+
+const metascraper = Metascraper([
+  require('metascraper-author')(),
+  require('metascraper-date')(),
+  require('metascraper-description')(),
+  require('metascraper-image')(),
+  require('metascraper-lang')(),
+  require('metascraper-lang-detector')(),
+  require('metascraper-logo')(),
+  // require('metascraper-clearbit-logo')(),
+  require('metascraper-publisher')(),
+  require('metascraper-title')(),
+  require('metascraper-url')(),
+  require('metascraper-audio')(),
+  require('metascraper-soundcloud')(),
+  require('metascraper-video')(),
+  require('metascraper-youtube')(),
+
+  // require('./rules/metascraper-embed')()
+])
+
+let oEmbedProvidersFile = fs.readFileSync(path.join(__dirname, './providers.json'), 'utf8')
+
+// some providers allow a format parameter
+// we need JSON
+oEmbedProvidersFile = oEmbedProvidersFile.replace('{format}', 'json')
+
+const oEmbedProviders = JSON.parse(oEmbedProvidersFile)
+
+const fetchEmbed = async url => {
+  const provider = oEmbedProviders.find(provider => {
+    return provider.provider_url.includes(url.hostname)
+  })
+  if (!provider) return {}
+  const {
+    endpoints: [endpoint],
+  } = provider
+  const endpointUrl = new URL(endpoint.url)
+  endpointUrl.searchParams.append('url', url.href)
+  endpointUrl.searchParams.append('format', 'json')
+  let json
+  try {
+    const response = await fetch(endpointUrl)
+    json = await response.json()
+  } catch (err) {
+    error(`Error fetching embed data: ${err.message}`)
+    return {}
+  }
+
+  return {
+    type: json.type,
+    html: json.html,
+    author: json.author_name,
+    date: json.upload_date,
+    sources: ['oembed'],
+  }
+}
+
+const fetchResource = async url => {
+  const response = await fetch(url)
+  const html = await response.text()
+  const resource = await metascraper({ html, url: url.href })
+  return {
+    sources: ['resource'],
+    ...resource,
+  }
+}
+
+export default async function scrape(url) {
+  url = new URL(url)
+  if (url.hostname === 'youtu.be') {
+    // replace youtu.be to get proper results
```


thank you for your strategic use of comments @roschaefer

File: backend/src/schema/resolvers/users.js [hunk 6]
```diff
@@ -36,16 +37,6 @@ const count = obj => {
...
-const undefinedToNull = list => {
-  const resolvers = {}
-  list.forEach(key => {
-    resolvers[key] = async (parent, params, context, resolveInfo) => {
-      return typeof parent[key] === 'undefined' ? null : parent[key]
-    }
-  })
-  return resolvers
-}
-
```
Comment (Tirokk) [inline backend/src/schema/resolvers/users.js:48]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-22T07:30:55Z" title="Monday, July 22nd 2019, 9:30:55 am +02:00">Jul 22, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -36,16 +37,6 @@ const count = obj => {
   return resolvers
 }
 
-const undefinedToNull = list => {
-  const resolvers = {}
-  list.forEach(key => {
-    resolvers[key] = async (parent, params, context, resolveInfo) => {
-      return typeof parent[key] === 'undefined' ? null : parent[key]
-    }
-  })
-  return resolvers
-}
-
```


Nice refactoring :smile_cat:

File: backend/src/schema/resolvers/embeds.spec.js (hunk 1 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/embeds.spec.js:204]: > [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-07-22T08:23:17Z" title="Monday, July 22nd 2019, 10:23:17 am +02:00">Jul 22, 2019</time>_

---


Is should be filmed when there is video or audio contend in an article as far as I understand metascraper.

File: backend/src/schema/resolvers/embeds/scraper.js (hunk 4 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/embeds/scraper.js:43]: > [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-07-22T08:27:35Z" title="Monday, July 22nd 2019, 10:27:35 am +02:00">Jul 22, 2019</time>_

---


oEmbed is a specefication which Sites can use for embeds but don’t have to. It’s a way I enhanced the metascraper results with data directly from the providers. There is an official oEmbed list which contains a lot of sites which supports that. Does that answer your question?

File: backend/src/schema/resolvers/embeds.spec.js (hunk 1 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/embeds.spec.js:204]: > [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-07-22T12:46:17Z" title="Monday, July 22nd 2019, 2:46:17 pm +02:00">Jul 22, 2019</time>_

---


*filled

**Review by Mogge — APPROVED** [review]
Hey @roschaefer Mr Rocket. 🚀

You did the next great step to have the embeds in our editor. Yuhuuu 💫

Big thanks for taking us forward … 😍

Elmo Dancing

**Pull request closed without merging**